### PR TITLE
samples: matter: implemented window covering application

### PR DESCRIPTION
**Pull request by markaj-nordic**
This application implements the Matter window covering device
which can be controlled with UI or remotely with Matter controller.

Signed-off-by: Marcin Kajor <marcin.kajor@nordicsemi.no>

### CODE REVIEW

File: samples/matter/common/src/pwm_device.cpp [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+#include <drivers/pwm.h>
+#include <logging/log.h>
+#include <zephyr.h>
+
+LOG_MODULE_DECLARE(app, CONFIG_MATTER_LOG_LEVEL);
+
+int PWMDevice::Init(const device * aPWMDevice, uint32_t aPWMChannel, uint8_t aMinLevel, uint8_t aMaxLevel, uint8_t aDefaultLevel)
+{
+    mState      = kState_On;
```
Comment (Damian-Nordic) [inline samples/matter/common/src/pwm_device.cpp:19]: Formatting is not compliant with NCS. We have .clang-format file in `samples/matter` so just make sure to reformat everything. E.g. you can use my script below to fix some of compliance issues.
```
#!/bin/bash

function all_files() {
  git ls-files . | grep -v '/zap-generated/'
}

function source_files() {
	all_files | grep -P '\.(h|cpp)$'
}

# fix file names
source_files | while read f; do
  dir="$(dirname "$f")"
  old_name="$(basename "$f")"
  new_name="$(echo "$old_name" | sed -E 's|(\w)([A-Z][a-z])|\1_\2|g' | tr '[A-Z]' '[a-z]')"
  [[ "$old_name" == "$new_name" ]] && continue
  git mv "$f" "$dir/$new_name"
  sed -i "s|$old_name|$new_name|g" $(all_files)
done

# fix comments

sed -i -E 's|//(.*)$|/*\1 */|g' $(source_files) # // -> /* */
clang-format --style=file -i $(source_files)
sed -i 's|2020 Nordic Semiconductor ASA|2021 Nordic Semiconductor ASA|g' $(all_files)
```

File: samples/matter/window_covering/sample.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    tags: ci_build
+    extra_args: CONF_FILE=prj_no_dfu.conf
+  matter.lock.low_power:
+    build_only: true
+    platform_allow: nrf52840dk_nrf52840 nrf5340dk_nrf5340_cpuapp
+    integration_platforms:
+      - nrf52840dk_nrf52840
+      - nrf5340dk_nrf5340_cpuapp
+    tags: ci_build
+    extra_args: "OVERLAY_CONFIG=../../overlay-low_power.conf"
```
Comment (kkasperczyk-no) [inline samples/matter/window_covering/sample.yaml:35]: overlay-low_power.conf is located in the window covering sample directory, so you don't need to add `../../`

File: samples/matter/window_covering/src/window-app.zap [hunk 3]
```diff
@@ -0,0 +1,9019 @@
...
+            }
+          ]
+        },
+        {
+          "name": "Window Covering",
+          "code": 258,
+          "mfgCode": null,
+          "define": "WINDOW_COVERING_CLUSTER",
+          "side": "server",
+          "enabled": 0,
```
Comment (kkasperczyk-no) [inline samples/matter/window_covering/src/window-app.zap:5086]: I wonder, why do we have Window Covering cluster disabled in Window Covering sample? Or am I missing something?

Comment (kkasperczyk-no) [reply samples/matter/window_covering/src/window-app.zap:5086]: Nevermind, it is enabled in other place :)

**Review by kkasperczyk-no — COMMENTED** [review]
This sample doesn't follow the new configuration style in our samples. Could you try align it with the following changes in other samples done in this PR: https://github.com/nrfconnect/sdk-nrf/pull/7497

File: samples/matter/window_covering/sample.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    tags: ci_build
+  matter.lock.release:
+    build_only: true
+    platform_allow: nrf52840dk_nrf52840 nrf5340dk_nrf5340_cpuapp
+    integration_platforms:
+      - nrf52840dk_nrf52840
+      - nrf5340dk_nrf5340_cpuapp
+    tags: ci_build
+    extra_args: CONF_FILE=prj_release.conf
+  matter.lock.no_dfu:
```
Comment (doublemis1) [inline samples/matter/window_covering/sample.yaml:20]: I Think that better option is to have matter.window_covering.both_dfu or something simmilar

File: samples/matter/window_covering/sample.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
+sample:
+  name: Matter Window Covering
+  description: Matter Window Covering example
+tests:
+  matter.lock.debug:
```
Comment (doublemis1) [inline samples/matter/window_covering/sample.yaml:5]: change for matter.window_covering.debug or matter/windcov.debug

File: samples/matter/window_covering/src/chip_project_config.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+ *          to the default configuration values for general CHIP features.
+ *
+ */
+
+#pragma once
+
+/* Use a default pairing code if one hasn't been provisioned in flash. */
+#define CHIP_DEVICE_CONFIG_USE_TEST_SETUP_PIN_CODE 20202021
+#define CHIP_DEVICE_CONFIG_USE_TEST_SETUP_DISCRIMINATOR 0xF00
+
```
Comment (doublemis1) [inline samples/matter/window_covering/src/chip_project_config.h:21]: Enable support for Extedned discovery
example:
#define CHIP_DEVICE_CONFIG_ENABLE_EXTENDED_DISCOVERY 1

File: samples/matter/window_covering/sample.yaml (hunk 2 above)
Comment (markaj-nordic) [reply samples/matter/window_covering/sample.yaml]: Done.

File: samples/matter/window_covering/src/chip_project_config.h (hunk 6 above)
Comment (markaj-nordic) [reply samples/matter/window_covering/src/chip_project_config.h:21]: Done

File: samples/matter/window_covering/sample.yaml (hunk 5 above)
Comment (markaj-nordic) [reply samples/matter/window_covering/sample.yaml]: Done

File: samples/matter/window_covering/sample.yaml (hunk 4 above)
Comment (markaj-nordic) [reply samples/matter/window_covering/sample.yaml]: The whole sample.yaml file was revised.

**Comment by markaj-nordic** [comment]
> This sample doesn't follow the new configuration style in our samples. Could you try align it with the following changes in other samples done in this PR: #7497

@kkasperczyk-no done

File: samples/matter/window_covering/src/app_task.cpp [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+
+	err = ThreadStackMgr().InitThreadStack();
+	if (err != CHIP_NO_ERROR) {
+		LOG_ERR("ThreadStackMgr().InitThreadStack() failed");
+		return err;
+	}
+
+#ifdef CONFIG_OPENTHREAD_MTD_SED
+	err = ConnectivityMgr().SetThreadDeviceType(ConnectivityManager::kThreadDeviceType_SleepyEndDevice);
+#else
```
Comment (kkasperczyk-no) [inline samples/matter/window_covering/src/app_task.cpp:95]:
```suggestion
#elif CONFIG_CHIP_THREAD_SSED
    err = ConnectivityMgr().SetThreadDeviceType(ConnectivityManager::kThreadDeviceType_SynchronizedSleepyEndDevice);
#else
```

**Review by kkasperczyk-no — APPROVED** [review]
I have only one minor suggestion. LGTM overall :+1:

File: samples/matter/window_covering/prj.conf [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+# PWM support
+CONFIG_PWM=y
+
+# OpenThread configs
+CONFIG_OPENTHREAD_NORDIC_LIBRARY_MTD=y
+CONFIG_OPENTHREAD_MTD=y
+CONFIG_OPENTHREAD_FTD=n
+
+# Default OpenThread network settings
+CONFIG_OPENTHREAD_PANID=4660
```
Comment (Damian-Nordic) [inline samples/matter/window_covering/prj.conf:30]: Unused, remove.

File: samples/matter/window_covering/sample.yaml [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      - nrf52840dk_nrf52840
+      - nrf5340dk_nrf5340_cpuapp
+    platform_allow: nrf52840dk_nrf52840 nrf5340dk_nrf5340_cpuapp
+    tags: ci_build
+  sample.matter.window_cover.no_dfu:
+    build_only: true
+    extra_args: CONF_FILE=prj_no_dfu.conf
+    integration_platforms:
+      - nrf21540dk_nrf52840
+    platform_allow: nrf21540dk_nrf52840
```
Comment (Damian-Nordic) [inline samples/matter/window_covering/sample.yaml:25]: I think nrf52 and nrf53 are also supported, so could be added here (but good that they're not included in integation_platforms).

**Review by Damian-Nordic — APPROVED** [review]

**Comment by greg-fer** [comment]
@markaj-nordic , pushed my changes in a commit, but had some merging-related adventures. Please check that everything is OK and I haven't messed up anything.

**Review by greg-fer — APPROVED** [review]
Approving, but please add changelog entries in Matter and Matter samples sections.

**Review by greg-fer — APPROVED** [review]

**Review by LuDuda — APPROVED** [review]

**Pull request merged**